README
=begin

= tools ディレクトリ内にある各スクリプトの説明

== TODO
最終的には
*説明のないもの/いらないものは削除。
*機能の類似したものは統合。
*使い勝手の統一。
の方向で。

==OmoiKondara
=== Summary
:開発者/メンテナ
:機能概要
 パッケージをビルド
:実行方法
example.OmoiKondaraを pkgs以下に.OmoiKondara という名前でコピー。
自分の環境に合わせて編集をする。
その後で、 
 OmoiKondara [オプション] [パッケージ名...]
:ライセンス
 Ruby
=== Known Bugs
 * get_no の後で cp_to_tree が呼び出されるが，cp_to_tree にバグがあり
   pkgs/hoge の直下に NoSource なファイルと同じ名前のファイルがあると，
   それが pkgs/hoge/SOURCES にコピーされてしまう [2003/12/09 muraken]．

==commit
:開発者/メンテナ
:機能概要
 簡易コミットツール。%changelogの最初のエントリがログとなる。
:実行方法
  commit パッケージ名...
:ライセンス

==environment.rb
:開発者/メンテナ
 crouton
:機能概要
 ~/.OmoiKondara(など)を読んで各種変数を設定
:実行方法
 Rubyの場合
  $:.unshfit(File.direname($0))
  require 'environment'
 Shellの場合
  PATH=`dirname $0`:$PATH
  eval `environment.rb`
:ライセンス
 Modified BSD

==listskip
:開発者/メンテナ
 crouton
:機能概要
 .SKIPのあるディレクトリをリスト
:実行方法
  listskip (引数なし)
:ライセンス
 Modified BSD

== speclint
:開発者/メンテナ
 crouton
:機能概要
 ガイドライン文書に従ったspecのチェック
:実行方法
  speclint [オプション] パッケージ名...
:ライセンス
 Modified BSD

== update-mph
:開発者/メンテナ
 crouton
:機能概要
 各ディレクトリにおいてmph-scanpackagesを実行
:実行方法
  update-mph [ディレクトリ...] (省略すると TOPDIR[-*]以下)
:ライセンス
 Modified BSD

== delobso.rb
:開発者/メンテナ
 dora, kazuhiko
:機能概要
 pkgs 以下をパースして、.OmoiKondara の TOPDIR 以下の不要なファイルを消す
:実行方法
 pkgs ディレクトリで実行する
:ライセンス
 ?

== uninstallobso.rb
:開発者/メンテナ
 zunda
:機能概要
 pkgs 以下をパースして、オプションで指定されたパッケージのバイナリパッケ
 ージ名を表示する。
:実行方法
 pkgs ディレクトリで実行する
:ライセンス
 see "uninstallobso.rb"

== listunused
:開発者/メンテナ
 kazuhiko
:機能概要
 SRPM に含まれないパッチなどを消す
:実行方法
 pkgs ディレクトリで
  listunused [SRPM ファイル...]
:ライセンス
 Ruby

== download_all_source
:開発者/メンテナ
 ?, kazuhiko
:機能概要
 ダウンロードが必要なファイルを全て取って来る
 .OmoiKondara の MIRROR には対応しているが URL_ALIAS には未対応
 TO.Nonfree や TO.Zoo のものも取って来るかどうかは、OmoiKondara と同じ
 ように -n や -m などで制御できる
:実行方法
 pkgs ディレクトリで
  download_all_source
:ライセンス
 ?

== pick_up_licenses
:開発者/メンテナ
  zunda
:機能概要
  引数で指定したソースツリーのファイルを走査してライセンスらしき文言を見つける。
:実行方法
    pick_up_licenses <ソースツリー>
  -vオプションを多くつけることによって、表示される情報が多くなる。
:ライセンス
  see "pick_up_licenses"

== touch_spec.rb
:開発者/メンテナ
  kazuhiko (kaz)
:機能概要
  同じ ver-rel の SRPM があれば、それに spec のタイムスタンプを合わせるスクリプト
:実行方法
  pkgs の中で ../tools/touch_spec.rb
  引き数はなし
:ライセンス
  Ruby

== correct-mtime
:開発者/メンテナ
  zunda
:機能概要
  ワーキングコピーの変更時刻をコミット時刻に合わせる
:実行方法
    correct-mtime [<ファイル/ディレクトリ>]
:ライセンス
  see "correct-mtime"

== addprops
:開発者/メンテナ
  tamo
:機能概要
  svn status で "?" と表示されるのがウザいので
  log などを無視させるために propset する
:実行方法
  cd pkgs; sh ../tools/addprops
:ライセンス
  Modified BSD

== mphdiff.rb
:開発者/メンテナ
  kazuhiko
:機能概要
  二つの mph ファイルの Depends および Provides を比較する
:実行方法
  mphdiff.rb old.mph new.mph
  (*.mphrpm ではなく *.mph を渡すこと)
:ライセンス
  Ruby

== listduplicate.rb
:開発者/メンテナ
  kazuhiko
:機能概要
  全バイナリパッケージから重複したファイルを見つける
:実行方法
  listduplicate.rb
:ライセンス
  Ruby

== updatespecdb
:開発者/メンテナ
  不明
:機能概要
  OmoiKondaraの下請けとして、specdbを更新する
:実行方法
  直接は実行されない
:ライセンス
  不明
  
== 内容テンプレート
:開発者/メンテナ
:機能概要
:実行方法
:ライセンス

== 説明のないファイル 
N/A

== 非スクリプトファイル
*ChangeLog
*Makefile
*rpmvercmp.c

== 削除したもの

必要であるならば README(このファイル) に説明を加えてから復活させること。

不要なものは積極的に消していこう。

*buildtree
*chgrel
*chgrel.STABLE
*chkcomps.rb
*ftpsearch
*getname.rb
*mkkondaraiso

== 説明が無いため、OBSOLETE送りにしたパッケージ
*FutoiMomonga
*OmoiKondara.test
*OmoiKondara2
*README.OmoiKondara2
*example.OmoiKondara2
*example.OmoiKondara.yaml
*omokon-modoki.rb
*README.omokon-modoki
*ana-rpm-size.rb
*checkspec.rb
*chkreq.rb
*chkrpm
*chkrpmdep
*chkrpmdep.db
*chkrpmdup
*chksource
*chksrpm
*cleanupsource
*cleanupsource.php
*depgraph
*deprpme
*gnu_md5sum_check.php
*gonome_rebuildagainst
*momodo-check.rb
*mphvercmp.rb
*myomokon.sh
*printrequires
*repo_check

== 使われないため、OBSOLETE送りにしたパッケージ
== OmoiKondara3.rb
:開発者/メンテナ
 muraken
:機能概要
 OmoiKondara で使用する予定のライブラリ．
:実行方法
 実行してはいけません．
:ライセンス
 Ruby

== tc_OmoiKondara.rb
:開発者/メンテナ
 muraken
:機能概要
 OmoiKondara3.rb のテストケース
:実行方法
 ruby tc_OmoiKondara.rb
:ライセンス
 Ruby

== lib/*
:開発者/メンテナ
 muraken
:機能概要
 様々なライブラリ
:実行方法
 実行しないでください
:ライセンス
 Ruby と同じ

レポジトリから消えていたファイル
==delobso==

== cpanflute
:開発者/メンテナ
 kazuhiko (もとは rpm-build パッケージのもの)
:機能概要
 CPAN module の tarball から spec ファイルを作る。
:実行方法
  cpanflute [tarball]
 とすると、/tmp/cpan/spec/ にディレクトリができるので、それを pkgs/ に持っ
 て来る。
:ライセンス
 GPL
:代替品
 cpanfluteの作成するspec fileは完成度が低いため、
 cpanspecパッケージを使うべし。

=end


ChangeLog
2008-04-19  Nishio Futoshi  <fut_nis@d3.dion.ne.jp>

	* addprops.ignore: add report.log.old

Sun Mar 30 08:45:35 2008  Nishio Futoshi  <fut_nis@d3.dion.ne.jp>

	* addprops.ignore: add report.log report.config

Mon Aug 29 11:01:46 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* delobso.rb: display missing files. add '-n', '-O' and '-L'
	options to display missing files in Nonfree, Orphan and Alter
	directories.

Fri May  6 20:53:57 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* delobso.rb: keep source files for other architectures.

Tue Apr 12 09:04:14 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* commit: check status of a spec file itself.

Sat Feb 26 16:03:38 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* OmoiKondara: add a new option '-i (--install)' to force install
	packages after build.

Wed Feb 23 09:38:11 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* OmoiKondara: use REMOVE* file only if binary packages are built.

Tue Feb 22 13:48:33 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* OmoiKondara: use environment.rb.

Tue Feb 22 10:50:27 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* environment.rb: provide $NOTFILE.
	* delobso.rb: use $NOTFILE. display missing files if $VERBOSE.

Thu Feb 17 10:17:08 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* OmoiKondara: new option '$LOG_FILE_COMPRESS' whose default value
	is true.
	(buildme): ignore 'SRPM.ONLY'.
	(backup_rpms): try 'rpm -Uvh' without '--nodeps' first.  update
	$SYSTEM_PROVIDES after 'rpm -Uvh'.

Tue Jan 25 11:39:24 2005  Kazuhiko  <kazuhiko@fdiary.net>

	* environment.rb: default architecture of ix86 is now i686.

	* OmoiKondara: default architecture of ix86 is now i686. NOT file
	for ix86 is now 'NOT.ix86'.

Mon Nov 22 01:00:08 2004  TAKAHASHI Tamotsu <tamo momonga-linux org>
	* speclint: check whether URL tag exists.
	We have too many specfiles with no URL tags.
	So speclint doesn't take them as errors.

Fri Nov 12 18:50:00 2004  TAKAHASHI Tamotsu <tamo momonga-linux org>
	* OmoiKondara: support cachecc1
	* OmoiKondara: modify English messages
	* OmoiKondara: print "not installed" if a BuildPreReq is missing
	* example.OmoiKondara: add USE_CACHECC1 and CACHECC1_DISTCCDIR

Fri Aug 20 22:38:19 2004  Kazuhiko  <kazuhiko@fdiary.net>
	* OmoiKondara: exit if local source is missing. exit if 'rpm
	--test' fails.

Tue Aug 17 18:36:00 2004  tab <tab@momonga-linux.org>

	* listunused: adapt TO.Orphan and TO.Alter

Mon Aug  9 23:54:35 2004  zunda <zunda at freeshell.org>

	* correct-mtime: added

Thu Jul  8 22:00:57 2004  zunda <zunda at freeshell.org>

	* OmoiKondara: hopefully better handling of macro expansion.

Fri Jul  2 00:26:21 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* commit: invoke 'cvs up' in each package's directory.

Sun May 30 16:11:48 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* updatespecdb: read './rpmrc' instead of '/usr/lib/rpmrc'.

Wed May  5 16:56:33 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* delobso.rb: add nil checks.

Wed Apr 28 11:14:12 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* download_all_source: add '-z, --zoo' option that is disabled by
	default.

Fri Apr 23 15:45:38 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* download_all_source: only recognize first 'TO.*' in alphabetical
	order

Fri Apr 23 14:11:17 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* OmoiKondara: only recognize first 'TO.*' in alphabetical order

Fri Apr 23 11:13:07 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* delobso.rb: only recognize first 'TO.*' in alphabetical order

Thu Apr 22 02:00:00 2004  Masayuki SANO <nosanosa@momonga-linux.org>

	* OmoiKondara: add -O option to build Orphan packages
	* OmoiKondara: add -L option to build Alter packages

Wed Apr  7 22:43:56 2004  Kazuhiko  <kazuhiko@fdiary.net>

	* listunused: revise for subversion

Sun Mar 28 15:45:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* OmoiKondara: always set --target $ARCHITECTURE

Sun Mar 14 01:40:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* commit: Zoo packages are not built by default
	* commit: add -z option to build Zoo packages
	* commit: obsolete -m option

Sun Mar 10 13:40:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* commit: fix listing files under version control

Sun Mar  6 13:40:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* commit: fix getting a list of versioning files from svn information
	
Sun Feb 29 13:40:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* touch_spec.rb: written by kazuhiko

Sun Feb 29 13:40:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* OmoiKondara: fix for subversion

Thu Jan 15 22:39:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* add '\' in rpmmacros for debug

Fri Jan  9 00:36:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* OmoiKondara: add option to let OmoiKondara ignore REMOVE* file

Sat Jan  3 18:32:00 2004  YAMAZAKI Makoto <zaki@zakky.org>

	* commit: refactoring
	* commit: commit all specs at once if svn

Sat Jan  3 14:34:06 2004  Kenta MURATA  <muraken2@nifty.com>

	* OmoiKondara: correct SKIP support.

Sat Jan  3 14:21:11 2004  Kenta MURATA  <muraken2@nifty.com>

	* OmoiKondara: support SKIP file.

Fri Jan  2 16:01:05 2004  Kenta MURATA  <muraken2@nifty.com>

	* lib/builtin-macros: new file describes rpm builtin macros.
	* lib/macro-syntax.bnf: new file describes rpm macro syntax by
	BNF.

Wed Dec 24 18:32:00 2003  YAMAZAKI Makoto <zaki@zakky.org>

	* commit: add svn support

Fri Dec 19 06:54:39 2003  Kenta MURATA  <muraken2@nifty.com>

	* lib/timeout.rb: stop default value for timeout.
	append read-only accessor for timeout.
	* lib/process.rb: ditto.
	* test/tc_timeout.rb: ditto.

Fri Dec 19 06:52:03 2003  Kenta MURATA  <muraken2@nifty.com>

	* lib/arrayof.rb: modify documentation.

	* lib/timeoutio.rb: separate from OmoiKondara3.rb, and rename
	TimeoutedIO to TimeoutIO.
	Supports writing methods.

	* lib/process.rb: separate from OmoiKondara3.rb, and rename
	TimeoutedProcess to TimeoutProcess.

	* OmoiKondara3.rb: TimeoutedIO separate to lib/timeoutio.rb.
	ChildProcess, PipedProcess and TimeoutedProcess separete to
	lib/process.rb.

	* test/tc_timeoutio.rb: separate from tc_OmoiKondara.rb.

environment.rb
#!/usr/bin/env ruby
# $Id: environment.rb,v 1.6 2003/10/22 14:47:37 kazuhiko Exp $

# $ rd2 -rrd/rd2man-lib <this_file> | nroff -man

=begin
= NAME
environment.rb - Sets up environment for momonga tools.
= SYNOPSIS
: In Ruby scripts
  At the beginning of the script, do
   $:.unshift File.dirname($0)
   load 'environment.rb'
: In Shell scripts
  At the beginning of the script, do
   PATH=`dirname $0`:$PATH
   eval `environment.rb`
= DESCRIPTION
((*environment.rb*)) sets up commonly used variables for Momonga's
building tools.

Currently, following variables are set. Symbols in left and right of a
slash('/') is for Shell scripts and Ruby scripts, respectively.

: (({ARCH})) / (({$ARCH}))
  Target architecture.
: (({TOPDIR})) / (({$TOPDIR}))
  Output top directory.
: (({PKGDIR})) / (({$PKGDIR}))
  Location of pkg CVS module.
: (({FTP_CMD})) / (({$FTP_CMD}))
  An external ftp command to be used.
: (({DISPLAY})) / (({$DISPLAY}))
  X server connection.
: (({MAILADDR})) / (({$MAILADDR}))
  Email address for anonymous ftp.
: (({DISTCC_VERBOSE})) / (({DISTCC_VERBOSE}))
  Distcc verbosity.
: (({NUMJOBS})) / (({$NUMJOBS}))
  Number of jobs.
: (({WORKDIR})) / (({$WORKDIR}))
  Working directory.
: ((*not available*)) / (({DISTCC_HOST}))
  List of distcc hosts.
: ((*not available*)) / (({URL_ALIAS}))
  List of URL rewriting rules.
: ((*not available*)) / (({$MIRROR}))
  List of mirror servers for source fetching.
Other variables will be added in the future.
= OPTIONS
None.
= AUTHOR
Written by OZAWA -Crouton- Sakuro <crouton@momonga-linux.org>
= BUG
Reports and enhancements are welcome.
=end
#'

$TOPDIR = nil
$PKGDIR = nil
$MIRROR = []
$DISTCC_HOSTS = []
$URL_ALIAS = {}

$CONF_FILES = %w(./.OmoiKondara ~/.OmoiKondara /etc/OmoiKondara.conf)

found_conf = false
$CONF_FILES.each do |conf|
  conf = File.expand_path(conf)
  next unless FileTest.exist?(conf)
  found_conf = true
  open(conf) do |f|
    f.each_line do |line|
      next  if line =~ /^#.*$/ or line =~ /^$/
      s = line.split
      v = s.shift
      v.upcase!
      case v
      when "TOPDIR"
	$TOPDIR = s.shift
      when "PKGDIR"
	$PKGDIR = s.shift
      when "MIRROR"
	while v = s.shift
	  $MIRROR << v
	end
      when "FTP_CMD"
	$FTP_CMD = s.join " "
      when "DISPLAY"
	$DISPLAY = s.join " "
      when "URL_ALIAS"
	$URL_ALIAS[Regexp.compile(s.first)] = s.last
      when "DISTCC_HOST"
	$DISTCC_HOSTS.push s.last if not $DISTCC_HOSTS.include?(s.last)
      when "DISTCC_VERBOSE"
	$DISTCC_VERBOSE = true
      when "NUMJOBS"
	$NUMJOBS = s.shift
      when "WORKDIR"
	$WORKDIR = s.shift
	if not File.directory?($WORKDIR) then
	  $stderr.puts "WARNING: invalid workdir. use default"
	  $WORKDIR = nil
	end
      end
    end
  end
  break
end

if found_conf == false
  $stderr.puts <<EOF
FATAL: .OmoiKondara not found.
F.Y.I: please copy tools/example.OmoiKondara to pkgs/.OmoiKondara
       and modify that.
EOF
  exit 1
end

%w(TOPDIR PKGDIR).each do |name|
  if eval('$' + name) == nil
    $stderr.puts "FATAL: Mandatory parameter #{name} is not set."
    exit 1
  end
end

$TOPDIR = File.expand_path $TOPDIR
$PKGDIR = File.expand_path $PKGDIR

$ARCH = `uname -m`.chomp

case $ARCH
when 'x86_64'
  $NOTFILE = 'NOT.x86_64'
  $ARCH = 'x86_64'
when /^i\d86$/
  $NOTFILE = 'NOT.ix86'
  $ARCH = 'i686'
when /^alpha/
  $NOTFILE = 'NOT.alpha'
  open('/proc/cpuinfo').readlines.each do |line|
    if line =~ /^cpu model\s*:\s*EV([0-9]).*$/ && $1 == '5'
      $ARCH = 'alphaev5'
      break
    end
  end
when 'mips'
  $NOTFILE = 'NOT.mips'
  open('/proc/cpuinfo').readlines.each do |line|
    if line =~ /^cpu model\s*:\s*R5900.*/
      $ARCH = 'mipsel'
      break
    end
  end
when /^ppc64/
  $NOTFILE = 'NOT.ppc64'
  $ARCH = 'ppc64'
when /^ppc/
  $NOTFILE = 'NOT.ppc'
  $ARCH = 'ppc'
when /^ia64/
  $NOTFILE = 'NOT.ia64'
  $ARCH = 'ia64'
else
  $stderr.puts %Q(WARNING: unsupported architecture #{$ARCH})
end

found_rpmvercmp = false
found_ftp_cmd = false
ftp_cmd_name = $FTP_CMD.split[0]
"#{ENV['PATH']}:../tools".split(/:/).each do |path|
  path = File.expand_path(path)
  if found_rpmvercmp == false && FileTest.exist?("#{path}/rpmvercmp")
    found_rpmvercmp = true
  end
  if found_ftp_cmd == false && FileTest.exist?("#{path}/#{ftp_cmd_name}")
    found_ftp_cmd = true
  end
end
if found_rpmvercmp == false
  $stderr.puts "FATAL: rpmvercmp not found."
  $stderr.puts "F.Y.I: please execute make command in directory 'tools'"
  exit 1
end
if found_ftp_cmd == false
  $stderr.puts "FATAL: #{ftp_cmd_name} not found."
  $stderr.puts "F.Y.I: please install #{ftp_cmd_name}."
  exit 1
end

if $0 == __FILE__
  %w(TOPDIR PKGDIR FTP_CMD DISPLAY MAILADDR ARCH DISTCC_VERBOSE NUMJOBS WORKDIR).each do |name|
    puts %Q(#{name}='#{eval('$' + name)}'; export #{name})
  end
end

